src/main/scala/io/iohk/ethereum/vm/OpCode.scala

```diff
@@ -0,0 +1,507 @@
+    JUMP,
+    JUMPI,
+    //PC,
+    //MSIZE,
+    //GAS,
+    JUMPDEST,
+
+    PUSH1,
+    PUSH2,
+    PUSH3,
+    PUSH4,
+    PUSH5,
+    PUSH6,
+    PUSH7,
+    PUSH8,
+    PUSH9,
+    PUSH10,
+    PUSH11,
+    PUSH12,
+    PUSH13,
+    PUSH14,
+    PUSH15,
+    PUSH16,
+    PUSH17,
+    PUSH18,
+    PUSH19,
+    PUSH20,
+    PUSH21,
+    PUSH22,
+    PUSH23,
+    PUSH24,
+    PUSH25,
+    PUSH26,
+    PUSH27,
+    PUSH28,
+    PUSH29,
+    PUSH30,
+    PUSH31,
+    PUSH32,
+
+    DUP1,
+    DUP2,
+    DUP3,
+    DUP4,
+    DUP5,
+    DUP6,
+    DUP7,
+    DUP8,
+    DUP9,
+    DUP10,
+    DUP11,
+    DUP12,
+    DUP13,
+    DUP14,
+    DUP15,
+    DUP16,
+
+    SWAP1,
+    SWAP2,
+    SWAP3,
+    SWAP4,
+    SWAP5,
+    SWAP6,
+    SWAP7,
+    SWAP8,
+    SWAP9,
+    SWAP10,
+    SWAP11,
+    SWAP12,
+    SWAP13,
+    SWAP14,
+    SWAP15,
+    SWAP16,
+
+    LOG0,
+    LOG1,
+    LOG2,
+    LOG3,
+    LOG4,
+
+    //CREATE,
+    //CALL,
+    //CALLCODE,
+    RETURN
+    //DELEGATECALL,
+    //SUICIDE
+  )
+
+  val byteToOpCode: Map[Byte, OpCode] =
+    opcodes.map(op => op.code -> op).toMap
+}
+
+/**
+  * Base class for all the opcodes of the EVM
+  *
+  * @param code Opcode byte representation
+  * @param delta number of words to be popped from stack
+  * @param alpha number of words to be pushed to stack
+  */
+sealed abstract class OpCode(val code: Byte, val delta: Int, val alpha: Int) {
+  def this(code: Int, pop: Int, push: Int) = this(code.toByte, pop, push)
+
+  def execute(state: ProgramState): ProgramState
+}
+
+case object STOP extends OpCode(0x00, 0, 0) {
+  def execute(state: ProgramState): ProgramState =
+    state.halt
+}
+
+sealed abstract class BinaryOp(code: Int)(val f: (DataWord, DataWord) => DataWord) extends OpCode(code.toByte, 2, 1) {
+  def execute(state: ProgramState): ProgramState = {
+    val updatedState = for {
+      popped <- state.stack.pop(2)
+      (Seq(a, b), stack1) = popped
+      res = f(a, b)
+      stack2 <- stack1.push(res)
+    } yield state.withStack(stack2).step()
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+sealed abstract class UnaryOp(code: Int)(val f: DataWord => DataWord) extends OpCode(code.toByte, 1, 1) {
+  def execute(state: ProgramState): ProgramState = {
+    val updatedState = for {
+      popped <- state.stack.pop
+      (a, stack1) = popped
+      res = f(a)
+      stack2 <- stack1.push(res)
+    } yield state.withStack(stack2).step()
+
+    updatedState.valueOr(state.withError)
+  }
+}
+
+case object ADD extends BinaryOp(0x01)(_ + _)
+
+case object MUL extends BinaryOp(0x02)(_ * _)
+
+case object SUB extends BinaryOp(0x03)(_ - _)
+
+case object DIV extends BinaryOp(0x04)((a, b) => if (b != 0) a / b else DataWord(0))
+
+case object EXP extends BinaryOp(0x0a)(_ ** _)
+
+case object LT extends BinaryOp(0x10)((a, b) => DataWord(a < b))
+
+case object EQ extends BinaryOp(0x14)((a, b) => DataWord(a == b))
+
+case object ISZERO extends UnaryOp(0x15)(a => DataWord(a == 0))
```

Comment:
So `a == 0` vs `a == DataWord.Zero` is a matter of conciseness/readability vs slight efficiency boost. I went for the former cause comparing `DataWord` to numbers feels natural to me.

Also with certain tweaks we could even write `UnaryOp(0x15)(_ == 0)` (or `UnaryOp(0x15)(_ == DataWord.Zero)`), but I'm not sure if it's worth the effort.

I think it boils down to a matter of taste, so I would appreciate a second opinion. Happy to change it if others feel the same way @LukasGasior1 @whysoserious